src/clone.c

```diff
@@ -959,19 +983,36 @@ FuncDeclaration *buildDtor(AggregateDeclaration *ad, Scope *sc)
         ex = new DotVarExp(loc, ex, v, 0);
         if (v->type->toBasetype()->ty == Tstruct)
         {
-            // this.v.__dtor()
+            // this.v.__xdtor()
+
+            // This is a hack so we can call destructors on const/immutable objects.
+            ex = new AddrExp(loc, ex);
+            ex = new CastExp(loc, ex, v->type->mutableOf()->pointerTo());
+            ex = new PtrExp(loc, ex);
+            if (stc & STCsafe)
+                stc = (stc & ~STCsafe) | STCtrusted;
+
             ex = new DotVarExp(loc, ex, sdv->dtor, 0);
             ex = new CallExp(loc, ex);
         }
         else
         {
-            // Typeinfo.destroy(cast(void*)&this.v);
```

Comment:
It's a simple matter of readability. Encoding all those implicit assumptions makes it much harder to understand, particularly for people less experienced with the compiler code.

```d
auto elemType = type->baseElemOf();
if (elemType != Tstruct)
    return;
if (type == Tstruct)
{
}
else
{
    // type has to be Tsarray b/c otherwise baseElemOf wouldn't be Tstruct
    // and baseElemOf does only peel of static arrays
}
```

vs.

```d
auto elemType = type->baseElemOf();
if (elemType != Tstruct)
    return;
switch (type)
{
case Tstruct:
    break;
case Tsarray:
    break;
default:
    assert(0);
}
```